text/0000-edition-based-method-disambiguation.md

```diff
@@ -0,0 +1,202 @@
+from the ambiguous method call syntax, `self.method(args...)` to the
+unambiguous function call syntax `Trait::method(self, args...)` is an easy
+change. In longer method call chains however there isn't a way to disambiguate
+the trait a method is associated with when calling it without splitting up that
+expression into multiple expressions, which can change drop behavior and
+prevent temporaries for living as long as they need to.
+
+This RFC intentionally avoids solving this problem or even proposing strawmen
+versions of the syntax to avoid distracting from the core issue, but at the
+same time it increases the need for a language syntax extension for quickly
+disambiguating the trait a method call should come from.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[This comment](https://github.com/rust-lang/rust/issues/88967#issuecomment-938024847) on the `Iterator::intersperse` issue details a few alternatives that were discussed by the libs team when we encountered that specific issue. These include both short term and long term suggestions.
+
+**Short Term**
+
+* use build scripts in crates where this breakage is expected to detect when
+  the breakage is introduced by the compiler and disable the conflicting APIs
+  internally using `#[cfg]` directives.
+* use `#[cfg(accessible)]` in the crate where the breakage is expected to
+  automatically remove the conflicting API when the upstream API is introduced.
+
+These solutions don't solve the problem generally, and instead address the
+specific breakage that are known or expected. We can and do catch many such
+issues via crater but we cannot test all crates via crater and we still end up
+breaking people's crates on nightly before the crater runs have had a chance to
+catch any breakage.
+
+**Longer Term**
+
+* encoding the released std version in crates published to crates.io when
+  uploading and ignore methods introduced in newer releases when building that
+  published dependency.
+* [Supertrait item shadowing](https://github.com/rust-lang/rfcs/pull/2845)
+
+These solutions also do not solve the problem generally. The former solution
+would only solve the problem for crates uploaded to crates.io, and the latter
+would only solve the problem for traits where the ambiguity is introduced by a
+supertrait.
+
+# Prior art
+[prior-art]: #prior-art
+
+- [previous discussion on irlo](https://internals.rust-lang.org/t/idea-paths-in-method-names/6834/14?u=scottmcm)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Interaction with autoderef
+
+We already have logic for preferring stable methods to unstable methods but it
+breaks[^1] if the auto deref for one requires one more level of indirection
+than the other. We should be careful to consider how autoderef behavior can
+affect edition based method disambiguation.
+
+[^1]: https://github.com/rust-lang/rust/issues/86682
+
+# Future possibilities
```

Comment:
In https://github.com/rust-lang/lang-team/blob/master/design-meeting-minutes/2022-02-16-libs-most-wanted.md#make-stable-and-unstable-available-for-third-party-crates it was mentioned that having `unstable`/`stable` available for ecosystem crates would be nice.

Would ecosystem crates want this kind of disambiguation support too?  Would editions be the right answer for crates to use?

Reply:
I would anticipate them potentially wanting something similar but I'd expect them to want it be based off of semver rather than editions so they can control their release cycle.

Reply:
I think that pushes me even more to the "this should use MSRV, not editions" position.

Because if this were based on the `regex = "1.5.4"` in a future world where it applies to crates, it seems like the closest parallel to that for the library is the `rust = "1.54"` property.

Then upgrading the toolchain doesn't pick up the new methods at all (or maybe it does with a deny lint, if there's no other option), and you start getting a "use UFCS for this call as it will become ambiguous in future" warning on calls that would do something different in an upgraded MSRV.

Coupling this to the edition just feels too indirect, to me.

Reply:
Updated the future possibilities section to mention this topic.